techmatters/flex-plugins
# CHI-1421: Custom contact components

## Description

* Support for custom components in forms:
* Adds a 'CustomContactComponent' form item type. This allows you to specify a custom component to include in your form - the type of component is specified using the name it was registered under (see below)
  * Adds a custom component registry to the plugin. This allows you to 'register' react components that you wish to use in your forms.
  * You register components by wrapping them in a generator function that passes in a taskSid or a contact ID iinto the component (the idea of the CustomContactComponent is that the component works with contact information, so expects either a task for new contacts or a contact for existing ones)
  * Adds a 'context' parameter to the form generation custom hook, allowing you to pass in contextual data like the contact or task ID associated with the form - this is so it can be pasdsed into the custom components
  * Adds an example component. This will eventually be the ReferralList, the first 'real' custom component, but is just a stub for now
*  Deduplicates missing definition version IDs before requesting the JSON from s3. Before there was a case if the case list loaded up several items with the same missing definition, duplicate requests for the same form definitions could be sent - not a huge issue but it's an easy optimisation to do.

Unit tests in #1272 

### Checklist
- N/A Feature flags added (covered under one big Resources flag

### Related Issues
Supports CHI-1421

### Verification steps
* Check the placeholder is present on Aselo Dev new case information tab forms but not for editing existing contacts
* Try adding this component to other contact forms, check it works
* Try adding it to the case editing forms, it should render an error because there is no contact / task ID in the context
